api/app/invoice/api.py

```diff
@@ -0,0 +1,355 @@
+}
+
+invoice_list_fields = {
+    'id': fields.Integer,
+    'customer_email': fields.String,
+    'customer_name': fields.String,
+    'client_reference_id': fields.String,
+    'iso_currency_code': fields.String,
+    'due_date': fields.DateTime(dt_format='iso8601'),
+    'created_by_user_id': fields.Integer,
+    'created_by_user': fields.String(attribute='created_by.full_name'),
+    'created_at': fields.DateTime(dt_format='iso8601'),
+    'total_amount': fields.Float,
+    'current_payment_status': fields.String(attribute='current_payment_status.payment_status')
+}
+
+invoice_payment_status_fields = {
+    'id': fields.Integer,
+    'payment_status': fields.String,
+    'created_at_unix': fields.Integer
+}
+
+class InvoiceAPI(InvoiceMixin, restful.Resource):
+    @auth_required
+    @marshal_with(invoice_fields)
+    def get(self):
+        args = self.req_parser.parse_args()
+        invoice_id = args['invoice_id']
+
+        user_id = g.current_user["id"]
+        current_user = user_repository.get_by_id(user_id)
+
+        invoice = invoice_repository.get_one_for_customer(invoice_id, current_user.email)
+        if invoice is None:
+            return INVOICE_NOT_FOUND
+
+        return invoice, 200
+
+class InvoicePaymentStatusApi(restful.Resource, InvoiceMixin):
+    @auth_required
+    @marshal_with(invoice_payment_status_fields)
+    def get(self):
+        args = self.req_parser.parse_args()
+
+        payment_status = invoice_repository.get_latest_payment_status(
+            args['invoice_id'],
+            g.current_user['email']
+        )
+
+        if not payment_status:
+            return INVOICE_NOT_FOUND
+
+        return payment_status, 200
+
+class InvoiceListAPI(restful.Resource):
+    @auth_required
+    @marshal_with(invoice_list_fields)
+    def get(self):
+        invoices = invoice_repository.get_all_for_user(g.current_user["id"])
+        return invoices, 200
+
+class InvoiceAdminAPI(InvoiceAdminMixin, restful.Resource):
+    @auth_required
+    @marshal_with(invoice_list_fields)
+    def get(self):
+        args = self.get_parser.parse_args()
+        event_id = args['event_id']
+
+        user_id = g.current_user["id"]
+        current_user = user_repository.get_by_id(user_id)
+
+        if not current_user.is_event_treasurer(event_id):
+            return FORBIDDEN
+
+        invoices = invoice_repository.get_for_event(event_id)
+        return invoices, 200
+
+    @auth_required
+    @marshal_with(invoice_list_fields)
+    def post(self):
+        args = self.post_parser.parse_args()
+        event_id = args['event_id']
+        due_date = args['due_date']
+
+        user_id = g.current_user["id"]
+        current_user = user_repository.get_by_id(user_id)
+
+        if not current_user.is_event_treasurer(event_id):
+            return FORBIDDEN
+        
+        if due_date < datetime.now():
+            return INVOICE_MUST_HAVE_FUTURE_DATE
+
+        offer_ids = args['offer_ids']
```

Comment:
It's not quite what's happening. We're generating a separate invoice for each offer (Id) passed in. The invoices will look identical though since we're using the same event fees (that's the list of event fee ids we pass in). The way I thought of this is that you can pass in all the offers that should have an invoice that have the same charges (for eg. Industry people you pass in the event fee if for the registration fee and then all the invoices generated will have that line item. For students you pass in the fee for the registration fee and the event fee id that credits them back so the invoice will have those two line items). I've also enforced rules so that an offer can't have more than one valid invoice and there should be a constraint as well on the one table that enforces that an invoice can only have one offer. Does that make sense?

Reply:
And the reason it's a list of offer ids is for convenience for the admin person. They can generate the invoices for a cohort all at once instead of one by one.